src/Common/src/System/Collections/Generic/ArrayBuilder.cs

```diff
@@ -0,0 +1,216 @@
+        /// </summary>
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(_array, _count);
+        }
+
+        /// <summary>
+        /// Returns an array with equivalent contents as this builder.
+        /// </summary>
+        /// <remarks>
+        /// Do not call this method twice on the same builder.
+        /// </remarks>
+        public T[] ToArray()
+        {
+            if (_count == 0)
+            {
+                return Array.Empty<T>();
+            }
+
+            Debug.Assert(_array != null); // Nonzero _count should imply this
+
+            T[] result = _array;
+            if (_count < result.Length)
+            {
+                // Avoid a bit of overhead (method call, some branches, extra codegen)
+                // which would be incurred by using Array.Resize
+                result = new T[_count];
+                Array.Copy(_array, 0, result, 0, _count);
+            }
+
+#if DEBUG
+            // Try to prevent callers from using the ArrayBuilder after ToArray, if _count != 0.
+            _count = -1;
+            _array = null;
+#endif
+
+            return result;
+        }
+
+        /// <summary>
+        /// Adds an item to the backing array, without checking if there is room.
+        /// </summary>
+        /// <param name="item">The item to add.</param>
+        /// <remarks>
+        /// Use this method if you know there is enough space in the <see cref="ArrayBuilder{T}"/>
+        /// for another item, and you are writing performance-sensitive code.
+        /// </remarks>
+        public void UncheckedAdd(T item)
+        {
+            Debug.Assert(_count < Capacity);
+
+            _array[_count++] = item;
+        }
+
+        /// <summary>
+        /// Adds new default-initialized slots to this <see cref="ArrayBuilder{T}"/>.
+        /// </summary>
+        /// <param name="count">The number of items to zero-extend by.</param>
+        /// <remarks>
+        /// Unless <see cref="Count"/> plus <paramref name="count"/> cannot fit into
+        /// <see cref="Capacity"/>, this method operates in O(1) time.
+        /// </remarks>
+        public void ZeroExtend(int count)
+        {
+            Debug.Assert(count >= 0);
+
+            int endCount = _count + count;
+            if (endCount > Capacity)
+            {
+                EnsureCapacity(endCount);
+            }
+            _count = endCount;
+        }
+
+        private void EnsureCapacity(int minimum)
+        {
+            Debug.Assert(minimum > Capacity);
+
+            int nextCapacity = 2 * Capacity + 1;
+
+            Debug.Assert(nextCapacity > 0); // Check for overflow.
+            nextCapacity = Math.Max(nextCapacity, minimum);
+            
+            // Array.Resize will unnecessarily copy the slots @ _count
+            // and after which are all default-initialized.
+            T[] next = new T[nextCapacity];
+            if (_count > 0)
```

Comment:
I wonder if this check actually hurts more than it helps (I've not measured).  If the majority case is going from 0 to 1 elements, then we avoid the nop call to Copy in the majority of the cases, but if the majority case is anything else, we have an unnecessary branch.